src/api/authereum/fastWithdrawal.js

```diff
@@ -0,0 +1,127 @@
+const {
+  fromQuantizedToBaseUnitsBN,
+  Joi,
+  toBN,
+  toQuantizedAmountBN
+} = require('dvf-utils')
+const { post } = require('request-promise')
+const calculateFact = require('../../lib/stark/calculateFact')
+const validateWithJoi = require('../../lib/validators/validateWithJoi')
+const DVFError = require('../../lib/dvf/DVFError')
+
+const address0 = '0x'.padEnd(42, '0')
+
+const getValidTokenInfo = dvf => token => {
+  const tokenInfo = dvf.token.getTokenInfoOrThrow(token)
+
+  if (!tokenInfo.starkVaultId) {
+    throw new DVFError(
+      'NO_STARK_VAULT_ID_FOR_TOKEN',
+      {token, context: 'fastWithdrawal'}
+    )
+  }
+
+  if (!tokenInfo.deversifiStarkVaultId) {
+    throw new DVFError(
+      'NO_DEVERSIFI_STARK_VAULT_ID_FOR_TOKEN',
+      {token, context: 'fastWithdrawal'}
+    )
+  }
+
+  return tokenInfo
+}
+
+const getFeeQuantised = async (dvf, token) => dvf
+  .fastWithdrawalFee(token)
+  .then(res => toBN(res.feeQuantised))
+
+const schema = Joi.object({
+  amount: Joi.amount().required(),
```

Comment:
I think since we use the `validateWithJoi` method, the `required()` is implicitly there already so everything is required unless it's optional

Reply:
there is a deeper problem here, as this PR is basically copy-paste + edits (similar to #188). This applies in general to how ledger and authereum integration is implemented on the client, and all of it should be refactored to remove the unnecessary code duplication by branching only where it's necessary (when signing, retrieving public key etc). And by branching I don't mean using `if/else` or `switch` but creating an appropriate abstraction layer around the different ways of transactions can be signed and injecting a concrete implementation of that abstract interface when client is initialised. I've designed such abstraction already and created a concrete implementation for [keystore](https://github.com/DeversiFi/dvf-client-js/blob/97202e994ae09183092a0575ed60e406191b6d31/src/lib/keystore/index.js) (which is used [here](https://github.com/DeversiFi/dvf-client-js/blob/97202e994ae09183092a0575ed60e406191b6d31/src/lib/dvf/createSignedTransaction.js#L11)). I believe that adding `authereum` or `ledger` integration should be as simple as implementing that abstraction for those providers.

Reply:
Thanks Adrian, I'll see if I can make the change

Reply:
@adrian-gierakowski Can you take a look at the updates I made to see if I'm going in the right direction? 
I've added a method `attachStarkProvider` to dvf initialization which will attach a provider if a `wallet` instance is passes 
Wallet has a type and meta field which can store necessary data (privateKey -> mm / keystore, path -> ledger, starkProvider -> authereum)
In case of metamask since we don't have the privateKey at the point of initialization (it needs to fetch getUserConfig first to see if it's a legacy wallet etc.), we attach a provider at a later point using  `dvf.util.attachStarkProvider`
So what we actually do is initialize dvf -> get user info -> attach the correct trading key
I've tested it with metamask and ledger and it works

Reply:
Is this `meta` field already existing or something new that will be set from the UI ?

Reply:
It will be something set from the UI